plugins/woocommerce/src/Internal/ProductQueryFilters/FilterDataProvider.php

```diff
@@ -0,0 +1,364 @@
+	 * @param array $query_vars The WP_Query arguments.
+	 * @return object
+	 */
+	public function get_filtered_price( $query_vars ) {
+		$pre_filter_data = $this->pre_get_filter_data( $query_vars, 'price' );
+
+		if ( ! empty( $pre_filter_data ) ) {
+			return $pre_filter_data;
+		}
+
+		$transient_key = $this->get_transient_key( $query_vars, 'price' );
+		$cached_data   = $this->get_cache( $transient_key );
+
+		if ( ! empty( $cached_data ) ) {
+			return $cached_data;
+		}
+
+		global $wpdb;
+
+		add_filter( 'posts_clauses', array( $this->filter_clauses_generator, 'add_query_clauses' ), 10, 2 );
+		add_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$query_vars['no_found_rows']  = true;
+		$query_vars['posts_per_page'] = -1;
+		$query_vars['fields']         = 'ids';
+		$query                        = new \WP_Query();
+
+		$query->query( $query_vars );
+		$product_query_sql = $query->request;
+
+		remove_filter( 'posts_clauses', array( $this->filter_clauses_generator, 'add_query_clauses' ), 10 );
+		remove_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$price_filter_sql = "
+		SELECT min( min_price ) as min_price, MAX( max_price ) as max_price
+		FROM {$wpdb->wc_product_meta_lookup}
+		WHERE product_id IN ( {$product_query_sql} )
+		";
+
+		$results = $wpdb->get_row( $price_filter_sql ); // phpcs:ignore
+
+		$this->set_cache( $transient_key, $results );
+
+		return $results;
+	}
+
+	/**
+	 * Get stock status counts for the current products.
+	 *
+	 * @param array $query_vars The WP_Query arguments.
+	 * @return array status=>count pairs.
+	 */
+	public function get_stock_status_counts( $query_vars ) {
+		$pre_filter_data = $this->pre_get_filter_data( $query_vars, 'stock' );
+
+		if ( ! empty( $pre_filter_data ) ) {
+			return $pre_filter_data;
+		}
+
+		$transient_key = $this->get_transient_key( $query_vars, 'stock' );
+		$cached_data   = $this->get_cache( $transient_key );
+
+		if ( ! empty( $cached_data ) ) {
+			return $cached_data;
+		}
+
+		global $wpdb;
+		$stock_status_options = array_map( 'esc_sql', array_keys( wc_get_product_stock_status_options() ) );
+
+		add_filter( 'posts_clauses', array( $this->filter_clauses_generator, 'add_query_clauses' ), 10, 2 );
+		add_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$query_vars['no_found_rows']  = true;
+		$query_vars['posts_per_page'] = -1;
+		$query_vars['fields']         = 'ids';
+		$query                        = new \WP_Query();
+
+		$query->query( $query_vars );
+		$product_query_sql = $query->request;
+
+		remove_filter( 'posts_clauses', array( $this->filter_clauses_generator, 'add_query_clauses' ), 10 );
+		remove_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$stock_status_counts = array();
+
+		foreach ( $stock_status_options as $status ) {
+			$stock_status_count_sql = $this->generate_stock_status_count_query( $status, $product_query_sql, $stock_status_options );
+
+			$result = $wpdb->get_row( $stock_status_count_sql ); // phpcs:ignore
```

Comment:
The statement on `$this->generate_stock_status_count_query` is not being prepared: to enhance security and prevent SQL injection vulnerabilities, especially in dynamic SQL generation, it's crucial to use prepared statements wherever possible.

Example:

```
$query = $wpdb->prepare("
        SELECT COUNT(DISTINCT posts.ID) as status_count
        FROM {$wpdb->posts} as posts
        INNER JOIN {$wpdb->postmeta} as postmeta ON posts.ID = postmeta.post_id
        AND postmeta.meta_key = '_stock_status'
        AND postmeta.meta_value = %s
        WHERE posts.ID IN ( {$product_query_sql} )", $status);
```

Reply:
I agree 💯 on security.

In my test, the query returns nothing if I use `prepare`. `$product_query_sql` is a prepared statement, looks like if we run it through another `prepare`, the query will be broken.

Also, correct me if I'm wrong, but I think the current one is secured:
- `$status` is escaped with `esc_sql()`.
- `$product_query_sql` is the request generated by `WP_Query`.

What do you think?